designs/dotnetcore-debugging.md

```diff
@@ -0,0 +1,196 @@
+# .NET Core 2.0 / 2.1 debugging
+
+This is a design for the .NET Core 2.0 and 2.1 missing debugging feature.
+
+### What is the problem
+
+Currently SAM CLI does not provide debugging support for .NET Core 2.0 and 2.1 because `dotnet` command does **not** provide _start and wait for the debugger_ out of the box and neither does current AWS Lambda runner has debugging support flag or anything. Apart from that .NET remote debugger ([vsdbg](https://aka.ms/getvsdbgsh)) could not attach via port, but instead only by PID of the running program from within the **target** machine (Docker container). To debug the app, customer can use `docker` pipe, and for that we need the exact container name (which must be the same for each run or predefined for the customer) to attach to, or some mechanism to provide piping support via proxy and expose the port (requires much more effort and additional investigation). 
+
+And let's break it down to the checklist
+
+* Put the `vsdbg` on the Docker container;
+* Add _"wait for the debugger"_ flag in runner program and implement corresponding functionality;
+* Implement corresponding container entry point override in SAM to support invoking the runner with this flag **on**.
+* Provide instructions for the customer of how to configure VS Code and VS to attach.
+
+### What will be changed
+
+Runner program for .NET core [2.0](https://github.com/lambci/docker-lambda/blob/master/dotnetcore2.0/run/MockBootstraps/Program.cs) and [2.1](https://github.com/lambci/docker-lambda/blob/master/dotnetcore2.1/run/MockBootstraps/Program.cs) in another [repo](https://github.com/lambci/docker-lambda) to support waiting for the debugger to attach. Already have [PR](https://github.com/lambci/docker-lambda/pull/127) (guys and @mhart - reviewers wanted :)). I think we should provide new `sam local` flag  `--container-name` which will assign desired name to the container with the Lambda runtime to simplify `docker exec` experience from VS Code later.
```

Comment:
Providing a container name is definitely something interesting to do, instead of auto generated names like `quirky_turtle` etc. I fully agree that docker exec becomes easier to do.

Reply:
Yes, I think is one can come very handy